### DISCUSSION TITLE
allowing custom serializer?

### DISCUSSION THREAD

**Discussion by arily** [nitrojs/nitro#1615] [Q&A]
Hi,
I am using nitro with Nuxt3.
as described in https://github.com/unjs/nitro/issues/804, we can only send json to the client.

with nuxt-trpc I am able provide my custom serializer like https://trpc.io/docs/server/data-transformers#different-transformers-for-upload-and-download shows, using `superjson` when client -> server and `uneval` in  server -> client.

with this approach I am serializing the response to string, then serialize this string to json again with json-rpc overheads.

I am looking for a way to optimize this output, a custom serializer in h3/nitro would be a perfect fit. 

is it possible to do something like I described?

**Comment by mubaidr** [comment]
Right now, if json/ object is returned from the nitro API, there is mismatch on client-side for types. An option to specify custom serializer would be awesome!

Just wondering we already have [@nuxt/devalue](https://github.com/nuxt-contrib/devalue#nuxtdevalue) module, wondering if the same can be used for this purpose?

**Comment by arily** [comment]
poc
`handler.ts`
```typescript
import { uneval } from 'devalue'
import type { EventHandler, EventHandlerRequest } from 'h3/dist/index'

export function defineUnevalEventHandler<TReq extends EventHandlerRequest, TRet>(fn: EventHandler<TReq, TRet>) {
  return defineEventHandler(async (event) => {
    try {
      const returnValue = await fn(event)
      event.node.res.end(uneval(returnValue))
    }
    catch (err: any) {
      event.node.res.writeHead(500)
      const returnValue = JSON.parse(JSON.stringify(err, Object.getOwnPropertyNames(err)))
      const { message = undefined, ...rest } = returnValue
      event.node.res.end(`throw new Error(${uneval(message)}, ${uneval(rest)})`)
    }
  }) as EventHandler<TReq, TRet>
}
```

`server/api/test-safe.ts`
```typescript
import { defineUnevalEventHandler } from '../handler'

export default defineUnevalEventHandler((event) => {
  return {
    article: 'hello world!',
    date: new Date(),
    set: new Set(),
    map: new Map([
      ['key', 'value'],
      ['key2', 'value2'],
    ]),
  }
})
```

`server/api/test-error.ts`
```typescript
import { defineUnevalEventHandler } from '../handler'

export default defineUnevalEventHandler((event) => {
  throw new Error('boom')
})

```

`GET /api/test-safe`
```javascript
{article:"hello world!",date:new Date(1699258213021),set:new Set([]),map:new Map([["key","value"],["key2","value2"]])}
```

`Get /api/test-error`
```javascript
throw new Error("boom", {
    stack: "Error: boom\n    at \u003Canonymous> ...(rest of the stack)"
})
```
